api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java

```diff
@@ -646,6 +655,88 @@ private static String sanitizeSimpleString(CharSequence value) {
     return String.format(Locale.ROOT, "(hash-%08x)", HASH_FUNC.apply(value));
   }
 
+  private static String sanitizeVariant(Variant value, long now, int today) {
+    return sanitizeVariant(value.value(), now, today);
+  }
+
+  private static String sanitizeVariant(VariantValue value, long now, int today) {
+    if (value instanceof VariantObject) {
+      return sanitizeVariantObject(value.asObject(), now, today);
+    } else if (value instanceof VariantPrimitive) {
+      return sanitizeVariantValue(value.asPrimitive(), value.type(), now, today);
+    } else {
+      return sanitizeVariantArray(value.asArray(), now, today);
+    }
+  }
+
+  private static String sanitizeVariantObject(VariantObject value, long now, int today) {
+    StringBuilder builder = new StringBuilder();
+    builder.append("{");
+    boolean first = true;
+    for (String field : value.fieldNames()) {
+      if (first) {
+        first = false;
+      } else {
+        builder.append(", ");
+      }
+      builder.append(String.format(Locale.ROOT, "(hash-%s)", field)).append(": ");
+      VariantValue fieldValue = value.get(field);
+      PhysicalType fieldType = fieldValue.type();
+      builder.append(sanitizeVariantValue(fieldValue, fieldType, now, today));
+    }
+    builder.append("}");
+    return builder.toString();
+  }
+
+  private static String sanitizeVariantArray(VariantArray value, long now, int today) {
+    StringBuilder builder = new StringBuilder();
+    builder.append("{");
+    boolean first = true;
+    for (int i = 0; i < value.numElements(); i++) {
+      if (first) {
+        first = false;
+      } else {
+        builder.append(", ");
+      }
+      builder.append(sanitizeVariantValue(value.get(i), value.get(i).type(), now, today));
+    }
+    builder.append("}");
+    return builder.toString();
+  }
+
+  private static String sanitizeVariantValue(
+      VariantValue fieldValue, PhysicalType fieldType, long now, int today) {
+    StringBuilder builder = new StringBuilder();
+    switch (fieldType) {
+      case INT8:
+      case INT16:
+      case INT32:
+      case INT64:
+      case FLOAT:
+      case DOUBLE:
+      case DECIMAL4:
+      case DECIMAL8:
+      case DECIMAL16:
+        builder.append(sanitizeNumber((Number) fieldValue.asPrimitive().get(), fieldType.name()));
+        break;
+      case DATE:
+        builder.append(sanitizeDate(((Number) fieldValue.asPrimitive().get()).intValue(), today));
+        break;
+      case TIMESTAMPTZ:
+      case TIMESTAMPNTZ:
```

Comment:
How about nano timestamp? I think we also need to handle them here as well. And also time type.